### PR TITLE
MM-39918: mmctl: list archived private channels in channel list command with system admin user

### PR DESCRIPTION
**Pull request by KuSh**
#### Summary
Add parameter in GetDeletedChannels to include private channels if required and enable it for users with PermissionManageSystem

I'm trying to add tests but I need a bit of time to understand how it works

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/19913
Jira https://mattermost.atlassian.net/browse/MM-39918

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/Makefile
Comment (hanzei) [inline server/Makefile:460]: Unrelated change

File: server/channels/api4/channel_test.go
Comment (hanzei) [inline server/channels/api4/channel_test.go:962]: :+1:

File: server/channels/store/storetest/channel_store.go
Comment (hanzei) [inline server/channels/store/storetest/channel_store.go:1]: I would love to see another test a new test that check the logic for private channels

File: server/channels/store/sqlstore/channel_store.go
Comment (hanzei) [inline server/channels/store/sqlstore/channel_store.go:1492]: `includePrivate` is not quite the right term, is it? Even non-admins can see private channels if they are a member. Is there a better term? Or can we add a comment explaining what the parameter does?

**Review by hanzei — CHANGES_REQUESTED** [review]
A few minor comments. Overall, great work! Thanks for the PR :rocket:

File: server/Makefile
Comment (KuSh) [reply server/Makefile:460]: Yes totally unrelated, there's another one in server/cmd/mmctl/commands/channel_test.go which is a typo fix but I thought there were not worth a dedicated PR. Should I split at least that one in another PR ?

File: server/channels/store/sqlstore/channel_store.go
Comment (KuSh) [reply server/channels/store/sqlstore/channel_store.go:1492]: that could be ignoreTeamMembership or has[or can]ManageSystem but I'm not good at naming things ;)

File: server/Makefile
Comment (hanzei) [reply server/Makefile:460]: Would you mind opening a separate PR for the typo fix and this one? That way, we keep the history clean.

Comment (KuSh) [reply server/Makefile:460]: Sure

Comment (KuSh) [reply server/Makefile]: Here it is https://github.com/mattermost/mattermost/pull/28710

File: server/channels/store/storetest/channel_store.go
Comment (KuSh) [reply server/channels/store/storetest/channel_store.go:1]: I've completed testChannelStoreGetDeleted tell me if that suffice

File: server/channels/store/sqlstore/channel_store.go
Comment (KuSh) [reply server/channels/store/sqlstore/channel_store.go]: Ok I've settled on skipTeamMembershipCheck tell me if that seems a better term

**Review by hanzei — CHANGES_REQUESTED** [review]
@KuSh The tests are failing. Can you please take a look?

```
=== Failed
=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL/GetDeleted (0.01s)
    channel_store.go:1067: 
        	Error Trace:	/mattermost/server/channels/store/storetest/channel_store.go:1067
        	            				/mattermost/server/channels/store/storetest/channel_store.go:87
        	Error:      	"[0xc002c470e0 0xc002c471d0]" should have 3 item(s), but has 2
        	Test:       	TestChannelStore/PostgreSQL/GetDeleted
        	Messages:   	wrong list length

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL (3.01s)

=== FAIL: channels/store/sqlstore TestChannelStore (3.01s)

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL/GetDeleted (re-run 1) (0.01s)
    channel_store.go:1067: 
        	Error Trace:	/mattermost/server/channels/store/storetest/channel_store.go:1067
        	            				/mattermost/server/channels/store/storetest/channel_store.go:87
        	Error:      	"[0xc00143f4a0 0xc00143f590]" should have 3 item(s), but has 2
        	Test:       	TestChannelStore/PostgreSQL/GetDeleted
        	Messages:   	wrong list length

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL (re-run 1) (0.01s)

=== FAIL: channels/store/sqlstore TestChannelStore (re-run 1) (0.01s)

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL/GetDeleted (re-run 2) (0.01s)
    channel_store.go:1067: 
        	Error Trace:	/mattermost/server/channels/store/storetest/channel_store.go:1067
        	            				/mattermost/server/channels/store/storetest/channel_store.go:87
        	Error:      	"[0xc001ed11d0 0xc001ed12c0]" should have 3 item(s), but has 2
        	Test:       	TestChannelStore/PostgreSQL/GetDeleted
        	Messages:   	wrong list length

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL (re-run 2) (0.01s)

=== FAIL: channels/store/sqlstore TestChannelStore (re-run 2) (0.01s)

=== FAIL: channels/store/sqlstore TestChannelStore/PostgreSQL/GetDeleted (re-run 3) (0.01s)
    channel_store.go:1067: 
        	Error Trace:	/mattermost/server/channels/store/storetest/channel_store.go:1067
        	            				/mattermost/server/channels/store/storetest/channel_store.go:87
        	Error:      	"[0xc001a0e0f0 0xc001a0e1e0]" should have 3 item(s), but has 2
        	Test:       	TestChannelStore/PostgreSQL/GetDeleted
        	Messages:   	wrong list length
```

**Comment by KuSh** [comment]
> @KuSh The tests are failing. Can you please take a look?

I can't reproduce locally :

```
❯ ./bin/gotestsum --packages="github.com/mattermost/mattermost/server/v8/channels/store/sqlstore"
✓  channels/store/sqlstore (7m36.817s)

=== Skipped
=== SKIP: channels/store/sqlstore TestFileInfoStore/MySQL/GetUptoNSizeFileTime (0.00s)
    file_info_store.go:873: MM-53905

=== SKIP: channels/store/sqlstore TestFileInfoStore/PostgreSQL/GetUptoNSizeFileTime (0.00s)
    file_info_store.go:873: MM-53905

=== SKIP: channels/store/sqlstore TestThreadStore/MySQL/ThreadStorePopulation/Get_unread_reply_counts_for_thread (0.00s)
    thread_store.go:421: MM-41797

=== SKIP: channels/store/sqlstore TestThreadStore/PostgreSQL/ThreadStorePopulation/Get_unread_reply_counts_for_thread (0.00s)
    thread_store.go:421: MM-41797

=== SKIP: channels/store/sqlstore TestUserStore/MySQL/GetProfilesNotInTeam/etag_for_profiles_not_in_team_1_after_u3_added_to_team_2 (0.00s)
    user_store.go:4790: 

=== SKIP: channels/store/sqlstore TestUserStore/PostgreSQL/GetProfilesNotInTeam/etag_for_profiles_not_in_team_1_after_u3_added_to_team_2 (0.00s)
    user_store.go:4790: 

DONE 5481 tests, 6 skipped in 493.378s

❯ ./bin/gotestsum --packages="github.com/mattermost/mattermost/server/v8/channels/store/storetest" -- -race $GOFLAGS -timeout=90m
✓  channels/store/storetest (2.087s)

DONE 1 tests in 14.031s
```

Does CI use a particular configuration ? Is it reproducible locally ?

**Comment by hanzei** [comment]
Hm, I can't reproduce the failure either. Let me merge `master` into your PR and see if that fixes the CI issue.

**Comment by KuSh** [comment]
> Hm, I can't reproduce the failure either. Let me merge `master` into your PR and see if that fixes the CI issue.

It doesn't seem to have helped :/
I'll try to follow exactly the steps the CI takes to see if I can reproduce locally. Meanwhile If you have any leads, I'm all ears!

**Comment by hanzei** [comment]
@KuSh You can run the tests locally using `go test -timeout 300s -run ^TestChannelStore$ ./channels/store/sqlstore -v`. With the command, they also error on my local setup.

**Comment by KuSh** [comment]
@hanzei please ignore my previous comment, I was missing the master merge you did. With that merge I can reproduce !

**Comment by KuSh** [comment]
@hanzei the problem should be fixed now

**Review by hanzei — APPROVED** [review]
Great work :rocket:

**Review by agarciamontoro — APPROVED** [review]
Code looks good to me! Thank you for your work :) What I'd love is for the Security team to take a further look just to double-check all is good, so let me request a review from them.

File: server/channels/store/sqlstore/channel_store.go
Comment (esarafianou) [inline server/channels/store/sqlstore/channel_store.go:1518]: A bit unrelated to the PR change, but since we're here, isn't that more clear?
```suggestion
                        SELECT * FROM Channels
                        WHERE 
                            (TeamId = ? OR TeamId = '')
                            AND DeleteAt != 0
                            AND (Type != ? 
                            OR (Type = ? AND Id IN (SELECT ChannelId FROM ChannelMembers WHERE UserId = ?)))
                        ORDER BY DisplayName LIMIT ? OFFSET ?
```

Comment (esarafianou) [reply server/channels/store/sqlstore/channel_store.go:1518]: We'd also need to change the args in line 1520.

**Review by esarafianou — APPROVED** [review]
LGTM. I have a non-security suggestion for the query construction. 

cc @KuSh, @agarciamontoro

File: server/channels/store/sqlstore/channel_store.go
Comment (KuSh) [reply server/channels/store/sqlstore/channel_store.go]: Great suggestion, it allowed me to use query builder !

**Comment by esarafianou** [comment]
@agarciamontoro since we changed the SQL query, can you re-review please?

**Comment by agarciamontoro** [comment]
Ooooh, that's a great improvement indeed, thanks! Let me double-check that the query plan is similar in both, will approve later if no performance issues comes up :)

**Comment by agarciamontoro** [comment]
We've ran an `EXPLAIN ANALYZE` on both the old and new queries on Community, using my user ID, and things seem to even be slightly better for the new query: the cost, the number of buffers hit, and the times are all lower in the new query :tada:

This is a specific example with a specific dataset, so things may change in other scenarios, but these results give me some peace of mind, I'm no longer concerned about potential performance regressions for the change. Thanks, @KuSh and @esarafianou, this looks good to me!

#### Old query
```
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|QUERY PLAN                                                                                                                                                                                                                                                                                                                                                                                                        |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|Limit  (cost=6732.67..6732.92 rows=100 width=1674) (actual time=18.186..18.202 rows=100 loops=1)                                                                                                                                                                                                                                                                                                                  |
|  Buffers: shared hit=6047                                                                                                                                                                                                                                                                                                                                                                                        |
|  ->  Sort  (cost=6732.67..6737.79 rows=2046 width=1674) (actual time=18.185..18.191 rows=100 loops=1)                                                                                                                                                                                                                                                                                                            |
|        Sort Key: channels.displayname                                                                                                                                                                                                                                                                                                                                                                            |
|        Sort Method: top-N heapsort  Memory: 100kB                                                                                                                                                                                                                                                                                                                                                                |
|        Buffers: shared hit=6047                                                                                                                                                                                                                                                                                                                                                                                  |
|        ->  HashAggregate  (cost=6634.02..6654.48 rows=2046 width=1674) (actual time=17.084..17.417 rows=955 loops=1)                                                                                                                                                                                                                                                                                             |
|              Group Key: channels.id, channels.createat, channels.updateat, channels.deleteat, channels.teamid, channels.type, channels.displayname, channels.name, channels.header, channels.purpose, channels.lastpostat, channels.totalmsgcount, channels.extraupdateat, channels.creatorid, channels.schemeid, channels.groupconstrained, channels.shared, channels.totalmsgcountroot, channels.lastrootpostat|
|              Batches: 1  Memory Usage: 612kB                                                                                                                                                                                                                                                                                                                                                                     |
|              Buffers: shared hit=6044                                                                                                                                                                                                                                                                                                                                                                            |
|              ->  Append  (cost=0.00..6536.83 rows=2046 width=1674) (actual time=0.010..16.130 rows=955 loops=1)                                                                                                                                                                                                                                                                                                  |
|                    Buffers: shared hit=6044                                                                                                                                                                                                                                                                                                                                                                      |
|                    ->  Seq Scan on channels  (cost=0.00..4561.94 rows=2045 width=186) (actual time=0.010..12.071 rows=906 loops=1)                                                                                                                                                                                                                                                                               |
|                          Filter: ((deleteat <> 0) AND (type <> 'P'::channel_type) AND (((teamid)::text = 'REDACTED'::text) OR ((teamid)::text = ''::text)))                                                                                                                                                                                                                                                      |
|                          Rows Removed by Filter: 83500                                                                                                                                                                                                                                                                                                                                                           |
|                          Buffers: shared hit=2876                                                                                                                                                                                                                                                                                                                                                                |
|                    ->  Nested Loop  (cost=0.84..1944.20 rows=1 width=186) (actual time=0.218..3.959 rows=49 loops=1)                                                                                                                                                                                                                                                                                             |
|                          Buffers: shared hit=3168                                                                                                                                                                                                                                                                                                                                                                |
|                          ->  Index Only Scan using idx_channelmembers_user_id_channel_id_last_viewed_at on channelmembers  (cost=0.42..240.74 rows=785 width=27) (actual time=0.041..0.448 rows=672 loops=1)                                                                                                                                                                                                     |
|                                Index Cond: (userid = 'REDACTED'::text)                                                                                                                                                                                                                                                                                                                                           |
|                                Heap Fetches: 474                                                                                                                                                                                                                                                                                                                                                                 |
|                                Buffers: shared hit=256                                                                                                                                                                                                                                                                                                                                                           |
|                          ->  Index Scan using channels_pkey on channels channels_1  (cost=0.42..2.17 rows=1 width=186) (actual time=0.005..0.005 rows=0 loops=672)                                                                                                                                                                                                                                               |
|                                Index Cond: ((id)::text = (channelmembers.channelid)::text)                                                                                                                                                                                                                                                                                                                       |
|                                Filter: ((deleteat <> 0) AND (type = 'P'::channel_type) AND (((teamid)::text = 'REDACTED'::text) OR ((teamid)::text = ''::text)))                                                                                                                                                                                                                                                 |
|                                Rows Removed by Filter: 1                                                                                                                                                                                                                                                                                                                                                         |
|                                Buffers: shared hit=2912                                                                                                                                                                                                                                                                                                                                                          |
|Planning:                                                                                                                                                                                                                                                                                                                                                                                                         |
|  Buffers: shared hit=553                                                                                                                                                                                                                                                                                                                                                                                         |
|Planning Time: 1.473 ms                                                                                                                                                                                                                                                                                                                                                                                           |
|Execution Time: 18.329 ms                                                                                                                                                                                                                                                                                                                                                                                         |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
#### New query
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|QUERY PLAN                                                                                                                                                                                               |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|Limit  (cost=5304.41..5304.66 rows=100 width=186) (actual time=12.343..12.359 rows=100 loops=1)                                                                                                          |
|  Buffers: shared hit=3132                                                                                                                                                                               |
|  ->  Sort  (cost=5304.41..5309.53 rows=2048 width=186) (actual time=12.342..12.349 rows=100 loops=1)                                                                                                    |
|        Sort Key: channels.displayname                                                                                                                                                                   |
|        Sort Method: top-N heapsort  Memory: 92kB                                                                                                                                                        |
|        Buffers: shared hit=3132                                                                                                                                                                         |
|        ->  Seq Scan on channels  (cost=242.71..5226.13 rows=2048 width=186) (actual time=0.017..11.682 rows=955 loops=1)                                                                                |
|              Filter: ((deleteat <> 0) AND (((teamid)::text = 'REDACTED'::text) OR ((teamid)::text = ''::text)) AND ((type <> 'P'::channel_type) OR ((type = 'P'::channel_type) AND (hashed SubPlan 1))))|
|              Rows Removed by Filter: 83451                                                                                                                                                              |
|              Buffers: shared hit=3132                                                                                                                                                                   |
|              SubPlan 1                                                                                                                                                                                  |
|                ->  Index Only Scan using idx_channelmembers_user_id_channel_id_last_viewed_at on channelmembers  (cost=0.42..240.74 rows=785 width=27) (actual time=0.032..0.429 rows=672 loops=1)      |
|                      Index Cond: (userid = REDACTED::text)                                                                                                                                              |
|                      Heap Fetches: 474                                                                                                                                                                  |
|                      Buffers: shared hit=256                                                                                                                                                            |
|Planning Time: 0.201 ms                                                                                                                                                                                  |
|Execution Time: 12.413 ms                                                                                                                                                                                |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

**Review by agarciamontoro — APPROVED** [review]

**Comment by hanzei** [comment]
@KuSh Could you please merge `master` into your branch? Thanks!

**Comment by KuSh** [comment]
@hanzei branch has been rebased!

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Confirmed I was experiencing the bug and deployed the solution
- Confirmed sys admin role users now see Archived private channels regardless of if they are members of the channel
- Regression tested that non-admin users see no change when using the `mcctl channel list` command
- General regression testing of `channel list`
No issues found 

LGTM!

**Comment by DHaussermann** [comment]
Huge thanks @KuSh for solving this 🎉

**Pull request merged**
